### ISSUE TITLE
`gap` doesn't work with `flex` on Safari

### ISSUE THREAD

**Issue by ludwiczakpawel** [forem/forem#12452]
**Describe the bug**

I've been living thinking that `gap` (as well as `grid-gap`) works perfectly fine with `flex` elements. Today I learned it's not supported in Safari which means there's a missing whitespace in Safari which I have already confirmed testing few views.

That means we have to fix all instances of a combination of two utility classes `flex` with `gap-*` in our views (`.html.erb`, `.js`, `.jsx`) AND also all instances of similar situation but in our SCSS files...

Yes, `gap` does work with `grid` but we should not simply replace every `flex` instance with `grid` because [_grid is for **layout** and flex is for **components**_](https://css-tricks.com/grid-for-layout-flexbox-for-components/).

So in most cases we will have to remove `gap` property and replace it with additional margin on child elements...

**To Reproduce**

Test our site in Safari.

**Desktop (please complete the following information):**

- Browser: Safari.

**Smartphone (please complete the following information):**

- Browser: Safari

**Comment by danieltott** [comment]
Definitely down to tackle this one! 

@ludwiczakpawel I was wondering how important the "grid is for layout and flex is for components" guideline is for forem. There are a number of instances of gap + flex in the codebase that would be a very simple swap to using grid instead of flex while not exposing it to any gotchas - especially leaning in to some `auto`-ish behaviors. 

For a small example:

```erb
# // app/views/admin/configs/show.html.erb
<fieldset class="flex flex-column gap-4">
```

This is a fieldset displaying `crayons-field` rows:

![image](https://user-images.githubusercontent.com/360261/106640787-7a648b80-6554-11eb-80b7-e795598d527a.png)

If we changed this to `<fieldset class="grid gap-4">` it behaves exactly the same and doesn't need any messing around with children and margins and `:last-child` shenanigans. 

I haven't looked at all of the spots where this occurs yet (streaming with @nickytonline tomorrow and want to _save some for the stream_ (is that a thing?)) - but I just wanted to see how strongly the component/layout guideline was being held. I personally think the use cases can be different for grid and flex - but I haven't found the breakdown to be easily split up like that. However this will be my first time contributing to forem so I figured I'd check in. 

This issue will really be a case-by-case thing, but _man_ will it save a lot of hassle and headache if we can avoid applying margins to children when possible.

Basically - is it cool if we use grid when appropriate even in 1D scenarios if we can make sure it won't cause any problems down the road?

**Comment by ludwiczakpawel** [comment]
> This issue will really be a case-by-case thing, but man will it save a lot of hassle and headache if we can avoid applying margins to children when possible.

100% agree. 

The component/layout guideline is more of a personal preference... I THINK grid may affect rendering performance when overused (but I don't have hard data to prove it though). 

Anyway, I guess for some bigger components that you think could be considered as _lightweight layouts_ (yes, that's very blurry explanation :D) we could just go ahead with classic grid, especially for cases when it would save tons of CSS related to `:last-child` and margins shenanigans. Since, as you mentioned, it's a case-by-case thing, I'd love to use flex whenever possible and whenever you think it won't require overengineering things. One example could be scenario where we have two buttons next to each other like

```
<div class="flex gap-2">
  <button />
  <button />
</div>
```

We could just replace `flex` with `grid` and call it a day, but in this case I think better approach would be:

```
<div class="flex">
  <button class="mr-2" />
  <button />
</div>
```

On the other hand, the fieldset example on your screenshot is probably good candidate to just go with `grid`.

Does that make some sense?

**Comment by danieltott** [comment]
@ludwiczakpawel Yep - that makes perfect sense! Just wanted to double check since this is my first time in this codebase. Thanks for the clarification!

As a possibly related question, how do we feel about js feature-checking for style stuff? Another approach could be to suss out when we're in the situation where flex+gap isn't supported and only apply fixes there, but unfortunately there's no way to get there with CSS-only solutions (`@supports` won't work here for example). I have in the past done this for very specific and small bits like this - kind of like getting very small pieces of [modernizr](https://modernizr.com/). The thought would be that a small function runs and adds a `no-flex-gap` class to the `body` or `html`, and then in the scss files we could do something like

```scss
.flex.gap-1 > * + .flex.gap-1 > * {
    margin-left: var(--su-1);
}
```

This code probably doesn't actually work as-is, but you can probably see where I'm going.

Pros:
- People writing markup / components could continue to author without worrying about browser support for this situation
- The `gap` property is one of the best things to come out of the last 5 years of CSS and it'd be a shame to not use it

Cons:
- Adds a small amount of overhead
- If there isn't already this sort of action happening, will take time to figure out where the best place to put it
- May be confusing to try to track down where that class is being added in the future
- It may be gross

To be absolutely clear, I am not even proposing this as an actual solution at this point - it may be total overkill for this. It just occurred to me at one point so I figured I'd see what the thinking is on this sort of approach _in general_.

**Comment by ludwiczakpawel** [comment]
re `@support` here's a read that could be useful (or not :D) https://ishadeed.com/article/flexbox-gap/

But generally... I think it's not worth the effort right now. Either with `@support` or some js solutions or css hacks - it's all gonna be a hack... If anything, I'd personally lean towards some `@support` solution because I guess it's the cleanest one. Although yea - whatever we'd go with here, I think it might be overkill.

The good news is: Safari Technology Preview does support `gap` for `flexbox` so I guess it'll come with one of the upcoming updates.

**Comment by danieltott** [comment]
Yep - was definitely using that article for some reference. I think that's probably the right call in general re: hacks. Thanks again for your time!

**Comment by danieltott** [comment]
Got a PR in for this one! https://github.com/forem/forem/pull/12667

**Comment by doublejosh** [comment]
Even grid gap doesn't work in Safari 14.0.3 😂
